### ISSUE TITLE
can resolved use avahi resolver? can resolved limit mdns to minimal ipv4?

### ISSUE THREAD

**Issue by xnox** [systemd/systemd#19434]
**Is your feature request related to a problem? Please describe.**
If mdns support is enabled in resolved, it seems to be enabled for both ipv4 & ipv6 and all domains.
If the host already has avahi installed and running, it doesn't seem to be used for resolutions.
Also some want to only configure ipv4, minimal, .local resolution in 169.254.x.x for mdns (unless that's what linklocal is?!)

**Describe the solution you'd like**
If avahi is running on the host, use it for ipv4 .local 169.254.x.x resolutions without having resolved bind to mdns port.
If avahi is not running allow resolved to only bind for ipv4, and only consider resolving .local 169.254.x.x addresses.

**Describe alternatives you've considered**
We have tried enabling resolved LinkLocal and MDNS but observed extremely long timeouts when trying to query non-existing domains. Since LinkLocal & mDNS take forever to respond to "no-existent.com".
Maybe we didn't configure it right?

**The systemd version you checked that didn't have the feature you are asking for**
247.3

**Comment by poettering** [comment]
I can't parse much of this. Is this a bug report?

> Since LinkLocal & mDNS take forever to respond to "no-existent.com".

Is "LinkLocal" supposed to mean LLMNR?

no-existent.com is neither a name suitable for LLMNR (as those are single-label names only) nor for mDNS (those have to end in .local). 

We added some fixes for mdns to v248 btw, please always test with that.

Is this a feature request to allow enabling of mDNS/LLMNR on ipv4 and on ipv6 independently of each other? what's the usecase?

**Comment by xnox** [comment]
> I can't parse much of this. Is this a bug report?
> 
> > Since LinkLocal & mDNS take forever to respond to "no-existent.com".
> 
> Is "LinkLocal" supposed to mean LLMNR?
> 

Yes, sorry.

> no-existent.com is neither a name suitable for LLMNR (as those are single-label names only) nor for mDNS (those have to end in .local). Which means the whole report doesn#t make much sense to me?
> 

hmm... so when glibc nss is talking to resolved over the stub resolver with hostnames a resolution for "nonexistent" will be tried by glibc's nss with all `search domains` as listed in the stub resolv.conf. I wonder, if in this case corporate & .local appear in /etc/resolv.conf for the end user, resulting in slow resolutions since "nonexistent" "nonexistent.corp" "nonexistent.local" are all tried and we wait for all of them to come back negative.


> We added some fixes for mdns to v248 btw, please always test with that.
> 

Will do.

> Is this a feature request to allow enabling of mDNS/LLMNR on ipv4 and on ipv6 independently of each other? what's the usecase?

Yes, as a parity for avahi's minimal mdnsv4 only. In Ubuntu, by default, we only enable minimal mdnsv4 and it would be nice to have that with systemd-resolved out of the box without needing to have extra libnss packages configured.

As far as I understand mDNSv4 is popular and widely deployed, whereas mDNSv6 not as much.

**Comment by pemensik** [comment]
Problem with avahi's implementation is it does not handle well resolution over both IPv4 and IPv6, in parallel. It should be improved first to enable it by default. Because the way it is implemented in nss-mdns, it tries first IPv4 and waits until it ends. Then it tries IPv6 and waits again. With ``mdns4_minimal`` nss plugin it takes 5 seconds, with ``mdns_minimal`` it takes 10 seconds. I have filled issue for that: https://github.com/lathiat/avahi/issues/421

The target is to make queries in parallel over both and shorten the search if found on one address family. So it works gracefully for hosts supporting only one AF, whichever it is.

Because resolved code is a fresh one, it might not suffer from similar limitation. I would be more happy to improve things in *avahi* instead of new implementation (bugs).
